File: public/logic/reducer.js
// im port { MOVE, START, ADD_TILE, UPDATE } from "/logic/actions.js";

// import { init } from "/game/init.js";
// import { addTile } from "/game/add.js";
// import { update } from "/game/tile.js";
// import { move } from "/game/move.js";

const reducer = (state, action) => {
  switch (action.type) {
    case "START": {
      let newState = {
        board: init(),
        changed: false
      };
      const { row, column, value } = chooseRandomTile(
        newState.board,
        action.randomPosition,
        action.randomValue
      );
      newState.board = addTile(newState.board, row, column, value);
      newState.board = update(newState.board);
      return newState;
    }
    case "MOVE": {
      let newState = move(state.board, action.direction);
      if (newState.changed) {
        const { row, column, value } = chooseRandomTile(
          newState.board,
          action.randomPosition,
          action.randomValue
        );
        newState.board = addTile(newState.board, row, column, value);
      }
      newState.board = update(newState.board);
      return newState;
    }
    default: {
      return state;
    }
  }
};

// export default reducer;

File: rust/src/reducer.rs
use state::State;
use action::Action;
use action::ActionType;

pub fn reducer(state: State, action: Action) -> State {
    match action.action_type {
        ActionType::Init => {
            let mut new_state = State::new();
            new_state.board.choose_random_tile(action.random_position.unwrap(), action.random_value.unwrap());
            new_state.board.update_tiles();
            new_state
        }
        ActionType::Move => {
            let mut new_state = state.clone();
            new_state.board.move_action(action.direction.unwrap());
            new_state.board.choose_random_tile(action.random_position.unwrap(), action.random_value.unwrap());
            new_state.board.update_tiles();
            new_state.won = new_state.board.has_won();
            new_state.lost = new_state.board.has_lost();
            new_state
        }
    }
}
